etcdserver/server_test.go

```diff
@@ -951,6 +951,149 @@ func TestSnapshot(t *testing.T) {
 	<-ch
 }
 
+// TestSnapshotOrdering ensures that when applying snapshot etcdserver renames snap db to db before raft persists snapshot to wal and snap files.
+func TestSnapshotOrdering(t *testing.T) {
+	n := newNopReadyNode()
+	st := store.New()
+	cl := membership.NewCluster("abc")
+	cl.SetStore(st)
+
+	testdir, err := ioutil.TempDir(os.TempDir(), "testsnapdir")
+	if err != nil {
+		t.Fatalf("Couldn't open tempdir (%v)", err)
+	}
+	defer os.RemoveAll(testdir)
+	if err := os.MkdirAll(testdir+"/member/snap", 0755); err != nil {
+		t.Fatalf("Couldn't make snap dir (%v)", err)
+	}
+
+	rs := raft.NewMemoryStorage()
+	p := mockstorage.NewStorageRecorderStream(testdir)
+	tr, snapDoneC := rafthttp.NewSnapTransporter(testdir)
+	r := newRaftNode(raftNodeConfig{
+		isIDRemoved: func(id uint64) bool { return cl.IsIDRemoved(types.ID(id)) },
+		Node:        n,
+		transport:   tr,
+		storage:     p,
+		raftStorage: rs,
+	})
+	s := &EtcdServer{
+		Cfg: &ServerConfig{
+			DataDir: testdir,
+		},
+		r:          *r,
+		store:      st,
+		cluster:    cl,
+		SyncTicker: &time.Ticker{},
+	}
+	s.applyV2 = &applierV2store{store: s.store, cluster: s.cluster}
+
+	be, tmpPath := backend.NewDefaultTmpBackend()
+	defer func() {
+		os.RemoveAll(tmpPath)
+	}()
+	s.kv = mvcc.New(be, &lease.FakeLessor{}, &s.consistIndex)
+	s.be = be
+
+	s.start()
+	defer s.Stop()
+
+	actionc := p.Chan()
+	n.readyc <- raft.Ready{Messages: []raftpb.Message{{Type: raftpb.MsgSnap}}}
+	ac := <-actionc
+	// MsgSnap triggers raftNode to call Save()
+	if ac.Name != "Save" {
+		t.Fatalf("expect save() is called, but got %v", ac.Name)
+	}
+	// get the snapshot sent by the transport
+	snapMsg := <-snapDoneC
+
+	// Snapshot ready triggers etcd server to rename snapshot db to db the first
+	// and raftnode to persist snapshot to wal and snap files the second.
+	snapMsg.Snapshot.Metadata.Index = 1
+	ready := raft.Ready{Snapshot: snapMsg.Snapshot}
+	n.readyc <- ready
+	var seenDBFilePath bool
+	timer := time.After(5 * time.Second)
+	for {
```

Comment:
the sequence can be `[Save(), DBFilePath(), SaveSnap()]` or  `[ DBFilePath(), Save(), SaveSnap()]`. all I care is that `DBFilePath()` happens before `SaveSnap()`